// tailwind-react-ui/plugins/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var visuallyHidden = (function (_temp) {
  var _ref = _temp === void 0 ? {} : _temp,
      _ref$variants = _ref.variants,
      variants = _ref$variants === void 0 ? ['responsive'] : _ref$variants;

  return function (_ref2) {
    var addUtilities = _ref2.addUtilities;
    addUtilities({
      '.visually-hidden, .visually-hidden-focusable': {
        border: 0,
        clip: 'rect(0 0 0 0)',
        height: '1px',
        margin: '-1px',
        overflow: 'hidden',
        padding: 0,
        position: 'absolute',
        width: '1px',
        whiteSpace: 'nowrap'
      },
      '.visually-hidden-focusable': {
        '&:active, &:focus, &:focus-within': {
          clip: 'auto',
          height: 'auto',
          margin: 0,
          overflow: 'visible',
          position: 'static',
          width: 'auto',
          whiteSpace: 'inherit'
        }
      }
    }, {
      variants: variants
    });
  };
});

exports.visuallyHidden = visuallyHidden;
